--- /dev/null
+++ ../src/Doctrine/Orm/Filter/SearchFilter.php
@@ -25,6 +25,7 @@
 use ApiPlatform\Metadata\Operation;
 use Doctrine\DBAL\Types\Types;
 use Doctrine\ORM\Query\Expr\Join;
+use Doctrine\ORM\Query\Parameter;
 use Doctrine\ORM\QueryBuilder;
 use Doctrine\Persistence\ManagerRegistry;
 use Psr\Log\LoggerInterface;
@@ -196,6 +197,7 @@
         }
 
         $metadata = $this->getNestedMetadata($resourceClass, $associations);
+        $doctrineTypeField = $this->getDoctrineFieldType($property, $resourceClass);
 
         if ($metadata->hasField($field)) {
             if ('id' === $field) {
@@ -211,7 +213,7 @@
                 return;
             }
 
-            $this->addWhereByStrategy($strategy, $queryBuilder, $queryNameGenerator, $alias, $field, $values, $caseSensitive);
+            $this->addWhereByStrategy($strategy, $queryBuilder, $queryNameGenerator, $alias, $field, $values, $caseSensitive, $doctrineTypeField);
 
             return;
         }
@@ -273,7 +275,7 @@
             $associationField = $associationFieldIdentifier;
         }
 
-        $this->addWhereByStrategy($strategy, $queryBuilder, $queryNameGenerator, $associationAlias, $associationField, $values, $caseSensitive);
+        $this->addWhereByStrategy($strategy, $queryBuilder, $queryNameGenerator, $associationAlias, $associationField, $values, $caseSensitive, $doctrineTypeField);
     }
 
     /**
@@ -281,7 +283,7 @@
      *
      * @throws InvalidArgumentException If strategy does not exist
      */
-    protected function addWhereByStrategy(string $strategy, QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $alias, string $field, mixed $values, bool $caseSensitive): void
+    protected function addWhereByStrategy(string $strategy, QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $alias, string $field, mixed $values, bool $caseSensitive, string $doctrineTypeField): void
     {
         if (!\is_array($values)) {
             $values = [$values];
@@ -295,14 +297,25 @@
             if (1 === \count($values)) {
                 $queryBuilder
                     ->andWhere($queryBuilder->expr()->eq($wrapCase($aliasedField), $wrapCase($valueParameter)))
-                    ->setParameter($valueParameter, $values[0]);
+                    ->setParameter($valueParameter, $values[0], $doctrineTypeField);
 
                 return;
             }
 
+            // get current parameters, because QueryBuilder->setParameters() erase previous parameters set
+            $parameters = $queryBuilder->getParameters();
+            $inQuery = [];
+            for ($i = 0; $i < \count($values); ++$i) {
+                $inQuery[] = $valueParameter;
+                $parameters->add(new Parameter($valueParameter, $caseSensitive ? $values[$i] : strtolower($values[$i]), $doctrineTypeField));
+                $valueParameter = ':'.$queryNameGenerator->generateParameterName($field);
+            }
+
+            // we cannot use expr->in() here because it considers $inQuery parameters as strings.
+
             $queryBuilder
-                ->andWhere($queryBuilder->expr()->in($wrapCase($aliasedField), $valueParameter))
-                ->setParameter($valueParameter, $caseSensitive ? $values : array_map('strtolower', $values));
+                ->andWhere($wrapCase($aliasedField).' IN ('.implode(', ', $inQuery).')')
+                ->setParameters($parameters);
 
             return;
         }
@@ -342,7 +355,7 @@
 
         $queryBuilder->andWhere($queryBuilder->expr()->orX(...$ors));
         foreach ($parameters as $parameter) {
-            $queryBuilder->setParameter($parameter[1], $parameter[0]);
+            $queryBuilder->setParameter($parameter[1], $parameter[0], $doctrineTypeField);
         }
     }
 
